Signal/src/call/PeerConnectionClient.swift

```diff
@@ -66,62 +66,134 @@ protocol PeerConnectionClientDelegate: class {
     func peerConnectionClient(_ peerconnectionClient: PeerConnectionClient, didUpdateRemote videoTrack: RTCVideoTrack?)
 }
 
-/**
- * `PeerConnectionClient` is our interface to WebRTC.
- *
- * It is primarily a wrapper around `RTCPeerConnection`, which is responsible for sending and receiving our call data 
- * including audio, video, and some post-connected signaling (hangup, add video)
- */
-class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelDelegate {
+// In Swift (at least in Swift v3.3), weak variables aren't thread safe. It
+// isn't safe to resolve/acquire/lock a weak reference into a strong reference
+// while the instance might be being deallocated on another thread.
+//
+// PeerConnectionProxy provides thread-safe access to a strong reference.
+// PeerConnectionClient has an PeerConnectionProxy to itself that its many async blocks
+// (which run on more than one thread) can use to safely try to acquire a strong
+// reference to the PeerConnectionClient. In ARC we'd normally, we'd avoid
+// having an instance retain a strong reference to itself to avoid retain
+// cycles, but it's safe in this case: PeerConnectionClient is owned (and only
+// used by) a single entity CallService and CallService always calls
+// [PeerConnectionClient terminate] when it is done with a PeerConnectionClient
+// instance, so terminate is a reliable place where we can break the retain cycle.
+//
+// Note that we use the proxy in two ways:
```

Comment:
The bad news is that upgrading to Swift 4.x doesn't fix these crashes.

The good news is that I figured out how to fix the remaining rare cases of this crash: by applying the same fix in more places.  Specifically, by using a delegate proxy that like `AtomicHandle` accesses the PCC in a safe way.  

So, I've renamed `AtomicHandle` to `PeerConnectionProxy` and we now use it two ways.